### ISSUE TITLE
ProcessPoolExecutor hangs when 1<max_tasks_per_child<num_submitted//max_workers

### ISSUE THREAD

**Issue by dalleyg** [python/cpython#115634]
# Bug report

### Bug description:

Starting in Python 3.11 when the `max_tasks_per_child` parameter was introduced, `ProcessPoolExecutor` hangs when `max_tasks_per_child>1` and enough tasks have been submitted to trigger a worker restart.

The following reproducer hangs on a fresh installation of Python 3.11 on Linux. 

```python
import os
from concurrent.futures import ProcessPoolExecutor
with ProcessPoolExecutor(1, max_tasks_per_child=2) as exe:
    futs = [exe.submit(os.getpid) for _ in range(10)]
    for fut in futs:
        print(fut.result())
```

Issuing a keyboard interrupt results in the following stack trace:

```
Traceback (most recent call last):
  File "<string>", line 7, in <module>
  File "/usr/lib/python3.11/concurrent/futures/_base.py", line 451, in result
    self._condition.wait(timeout)
  File "/usr/lib/python3.11/threading.py", line 320, in wait
    waiter.acquire()
KeyboardInterrupt

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "<string>", line 4, in <module>
  File "/usr/lib/python3.11/concurrent/futures/_base.py", line 647, in __exit__
    self.shutdown(wait=True)
  File "/usr/lib/python3.11/concurrent/futures/process.py", line 825, in shutdown
    self._executor_manager_thread.join()
  File "/usr/lib/python3.11/threading.py", line 1112, in join
    self._wait_for_tstate_lock()
  File "/usr/lib/python3.11/threading.py", line 1132, in _wait_for_tstate_lock
    if lock.acquire(block, timeout):
       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
KeyboardInterrupt
```

Notes:
- Interleaving submissions and result checking does not help.
- Adding a timeout to the `result()` calls does not help.
- Increasing the pool size does not help, unless it's made large enough to not require any worker restarts.
- Interestingly, setting `max_tasks_per_child=1` works great. It never hangs, and a new process correctly used for each task.
- It does not hang if `max_tasks_per_child` is set high enough so that no worker restarts happen.
- I have reproduced this problem in the following test environments:
  - On GitHub's default Linux CI environment using Python 3.11.
  - On GitHub's default Linux CI environment using Python 3.12.
  - On Ubuntu 22.04.3 LTS running under WSL2 using a fresh installation of Python 3.11.

### CPython versions tested on:

3.11, 3.12

### Operating systems tested on:

Linux

### Linked PRs
* gh-115642
* gh-140897
* gh-140900
* gh-143302
* gh-143303

**Comment by gaogaotiantian** [comment]
Can't repro this on WSL2 + 3.12/3.11/main. Result:

```
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/usr/lib/python3.12/multiprocessing/spawn.py", line 122, in spawn_main
    exitcode = _main(fd, parent_sentinel)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.12/multiprocessing/spawn.py", line 131, in _main
    prepare(preparation_data)
  File "/usr/lib/python3.12/multiprocessing/spawn.py", line 246, in prepare
    _fixup_main_from_path(data['init_main_from_path'])
  File "/usr/lib/python3.12/multiprocessing/spawn.py", line 297, in _fixup_main_from_path
    main_content = runpy.run_path(main_path,
                   ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "<frozen runpy>", line 286, in run_path
  File "<frozen runpy>", line 98, in _run_module_code
  File "<frozen runpy>", line 88, in _run_code
  File "/home/gaogaotiantian/programs/mycpython/example.py", line 4, in <module>
    futs = [exe.submit(os.getpid) for _ in range(10)]
            ^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.12/concurrent/futures/process.py", line 835, in submit
    self._adjust_process_count()
  File "/usr/lib/python3.12/concurrent/futures/process.py", line 794, in _adjust_process_count
    self._spawn_process()
  File "/usr/lib/python3.12/concurrent/futures/process.py", line 812, in _spawn_process
    p.start()
  File "/usr/lib/python3.12/multiprocessing/process.py", line 121, in start
    self._popen = self._Popen(self)
                  ^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.12/multiprocessing/context.py", line 289, in _Popen
    return Popen(process_obj)
           ^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.12/multiprocessing/popen_spawn_posix.py", line 32, in __init__
    super().__init__(process_obj)
  File "/usr/lib/python3.12/multiprocessing/popen_fork.py", line 19, in __init__
    self._launch(process_obj)
  File "/usr/lib/python3.12/multiprocessing/popen_spawn_posix.py", line 42, in _launch
    prep_data = spawn.get_preparation_data(process_obj._name)
                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.12/multiprocessing/spawn.py", line 164, in get_preparation_data
    _check_not_importing_main()
  File "/usr/lib/python3.12/multiprocessing/spawn.py", line 140, in _check_not_importing_main
    raise RuntimeError('''
RuntimeError: 
        An attempt has been made to start a new process before the
        current process has finished its bootstrapping phase.

        This probably means that you are not using fork to start your
        child processes and you have forgotten to use the proper idiom
        in the main module:

            if __name__ == '__main__':
                freeze_support()
                ...

        The "freeze_support()" line can be omitted if the program
        is not going to be frozen to produce an executable.

        To fix this issue, refer to the "Safe importing of main module"
        section in https://docs.python.org/3/library/multiprocessing.html
        
Traceback (most recent call last):
  File "/home/gaogaotiantian/programs/mycpython/example.py", line 6, in <module>
    print(fut.result())
          ^^^^^^^^^^^^
  File "/usr/lib/python3.12/concurrent/futures/_base.py", line 456, in result
    return self.__get_result()
           ^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.12/concurrent/futures/_base.py", line 401, in __get_result
    raise self._exception
concurrent.futures.process.BrokenProcessPool: A process in the process pool was terminated abruptly while the future was running or pending.
```

**Comment by brianschubert** [comment]
Can't reproduce on Linux + 3.12.1/3.11.7/main. I get the same `BrokenProcessPool` as @gaogaotiantian, even when `max_tasks_per_child=1`.

**Comment by dalleyg** [comment]
Presumably getting `BrokenProcessPool` is also a bug; just a different one (and a better one). Or am I misunderstanding?

Here are the specific Python version I tested on WSL2 + Ubuntu. It was installed with `apt`:
```
% python3.11 --version
Python 3.11.0rc1
```

**Comment by brianschubert** [comment]
@dalleyg `3.11.0rc1` is a pre-release version of Python 3.11. Are you able to reproduce this with a "full" release version of Python? Preferably one of the latest minor releases, currently 3.12.2 and 3.11.8.

**Comment by dalleyg** [comment]
I just downloaded the 3.11.8 tarball from python.org, ran `./configure` and `make`, moved my `~/.local` directory out of the way, and I still get the same hanging issue.

```
~/Python-3.11.8% ./python -c '
import os
from concurrent.futures import ProcessPoolExecutor
with ProcessPoolExecutor(1, max_tasks_per_child=2) as exe:
    futs = [exe.submit(os.getpid) for _ in range(10)]
    for fut in futs:
        print(fut.result())
'
139562
139562
^C^CTraceback (most recent call last):
  File "<string>", line 7, in <module>
  File "/home/dalleyg/projects/Python-3.11.8/Lib/concurrent/futures/_base.py", line 451, in result
    self._condition.wait(timeout)
  File "/home/dalleyg/projects/Python-3.11.8/Lib/threading.py", line 327, in wait
    waiter.acquire()
KeyboardInterrupt

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "<string>", line 4, in <module>
  File "/home/dalleyg/projects/Python-3.11.8/Lib/concurrent/futures/_base.py", line 647, in __exit__
    self.shutdown(wait=True)
  File "/home/dalleyg/projects/Python-3.11.8/Lib/concurrent/futures/process.py", line 851, in shutdown
    self._executor_manager_thread.join()
  File "/home/dalleyg/projects/Python-3.11.8/Lib/threading.py", line 1119, in join
    self._wait_for_tstate_lock()
  File "/home/dalleyg/projects/Python-3.11.8/Lib/threading.py", line 1139, in _wait_for_tstate_lock
    if lock.acquire(block, timeout):
       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
KeyboardInterrupt
```
I get the same hanging result on WSL2 + Ubuntu with:
* The official Ubuntu Python 3.11 and 3.12 packages (detailed in previous posts).
* Building Python 3.11.8 from scratch (immediately above).
* Building Python 3.12.2 from scratch (similar to above).
* Installing Python 3.11.7 from ppa:deadsnakes (one subminor back version: the latest available).
* Installing Python 3.12.1 from ppa:deadsnakes (one subminor back version: the latest available).

On the other hand, a colleague is seeing the `BrokenProcessPool` exception that both of you are getting on:
* Bare Metal Linux with Python 3.11.7
* Bare Metal Linux with Python 3.11.8
* Bare Metal Linux with Python 3.12.2
* The Docker Python 3.12.2 image from https://hub.docker.com/_/python/

This makes me wonder if the hanging bug is WSL2 and/or Ubuntu-specific, and the `BrokenProcessPool` bug instead occurs on native Linux systems.

Do you recommend that I fork this issue into two?
* Edit this issue to be WSL2-specific.
* Create a new issue for the `BrokenProcessPool` problem.

**Comment by dalleyg** [comment]
If I use Docker, I get the `BrokenProcessPool` exception as well:

```
% cat issue115634.py
import os
from concurrent.futures import ProcessPoolExecutor

with ProcessPoolExecutor(1, max_tasks_per_child=2) as exe:
    futs = [exe.submit(os.getpid) for _ in range(10)]
    for fut in futs:
        print(fut.result())

% docker run -it --rm --name 115634 -v "$PWD":/usr/src/myapp -w /usr/src/myapp python:3.11 python3 --version
Emulate Docker CLI using podman. Create /etc/containers/nodocker to quiet msg.
Python 3.11.8

% docker run -it --rm --name 115634 -v "$PWD":/usr/src/myapp -w /usr/src/myapp python:3.11 python issue115634.py
...SNIP...
concurrent.futures.process.BrokenProcessPool: A process in the process pool was terminated abruptly while the future was running or pending.
```

**Comment by gaogaotiantian** [comment]
I failed to repro this on WSL2 + Ubuntu 20.04 so I don't believe WSL is the sufficient trigger.

**Comment by dalleyg** [comment]
Since others can't reproduce my hanging problem, I'll close this issue for now and open a new one for the `BrokenProcessPool` error. 

Thanks for the help!

**Comment by dalleyg** [comment]
Okay, I have a reproducer now, so I'm reopening this. 

When running the code as a script, `BrokenProcessPool` is the result. When running it with `python -c`, hanging is the result. Here's a `bash` reproducer for both using Docker images (note: I'm using some syntax here that's not compatible with `zsh` so be sure to use `bash`):

```bash
echo '
import os
from concurrent.futures import ProcessPoolExecutor
with ProcessPoolExecutor(1, max_tasks_per_child=2) as exe:
    futs = [exe.submit(os.getpid) for _ in range(10)]
    for fut in futs:
        print(fut.result())
' > broken.py

(
    echo "TESTING WITH A SCRIPT"
    for v in 3.10 3.11.8 3.12.2 3.13.0a4; do
        DOCKER="docker run -it --rm --name 115634 -q -v "$PWD":/usr/src/myapp -w /usr/src/myapp python:$v"
        echo "--------------------------------------"
        (
            $DOCKER python3 --version &&
                $DOCKER timeout --signal=SIGINT --kill-after 5s -v 2s python3 broken.py
        ) 2>&1
    done
) | tee broken-script.log

(
    echo "TESTING WITH THE SCRIPT INLINED INTO THE COMMAND"
    for v in 3.10 3.11.8 3.12.2 3.13.0a4; do
        DOCKER="docker run -it --rm --name 115634 -q -v "$PWD":/usr/src/myapp -w /usr/src/myapp python:$v"
        echo "--------------------------------------"
        (
            $DOCKER python3 --version &&
                $DOCKER timeout --signal=SIGINT --kill-after 5s -v 2s python3 -c "$(cat broken.py)"
        ) 2>&1
    done
) | tee broken-inline.log
```

Here's a snipped version of `broken-script.log` that shows the eager exceptions. The full tracebacks look like the one shown in https://github.com/python/cpython/issues/115634#issuecomment-1951425366.
```
% cat broken-script.log
TESTING WITH A SCRIPT
--------------------------------------
Python 3.10.13
Traceback (most recent call last):
  File "/usr/src/myapp/broken.py", line 4, in <module>
    with ProcessPoolExecutor(1, max_tasks_per_child=2) as exe:
TypeError: ProcessPoolExecutor.__init__() got an unexpected keyword argument 'max_tasks_per_child'
--------------------------------------
Python 3.11.8
...SNIP...
concurrent.futures.process.BrokenProcessPool: A process in the process pool was terminated abruptly while the future was running or pending.
--------------------------------------
Python 3.12.2
...SNIP...
concurrent.futures.process.BrokenProcessPool: A process in the process pool was terminated abruptly while the future was running or pending.
--------------------------------------
Python 3.13.0a4
...SNIP...
concurrent.futures.process.BrokenProcessPool: A process in the process pool was terminated abruptly while the future was running or pending.
```

And here's what I see in `broken-inline.log` (complete output):
```
cat broken-inline.log
TESTING WITH THE SCRIPT INLINED INTO THE COMMAND
--------------------------------------
Python 3.10.13
Traceback (most recent call last):
  File "<string>", line 4, in <module>
TypeError: ProcessPoolExecutor.__init__() got an unexpected keyword argument 'max_tasks_per_child'
--------------------------------------
Python 3.11.8
4
4
timeout: sending signal INT to command ‘python3’
timeout: sending signal KILL to command ‘python3’
--------------------------------------
Python 3.12.2
4
4
timeout: sending signal INT to command ‘python3’
timeout: sending signal KILL to command ‘python3’
--------------------------------------
Python 3.13.0a4
4
4
timeout: sending signal INT to command ‘python3’
timeout: sending signal KILL to command ‘python3’
```

**Comment by gaogaotiantian** [comment]
I'm able to repro with
```python
import os
from concurrent.futures import ProcessPoolExecutor
if __name__ == '__main__':
    with ProcessPoolExecutor(1, max_tasks_per_child=2) as exe:
        futs = [exe.submit(os.getpid) for _ in range(10)]
        for fut in futs:
            print(fut.result())
```

**Comment by dalleyg** [comment]
Confirmed. With the `__name__` guard, I get the hanging for both the script and the inlined script.

**Comment by gaogaotiantian** [comment]
There's something fundamentally wrong here for the process numbers, I'll work on it.

**Comment by gaogaotiantian** [comment]
See #115642 for the detailed explanation and the potential fix.

**Comment by dalleyg** [comment]
That was quick! I'll watch that PR for a resolution.

In the meantime, I've put a safety check in my code. I'll adjust it once there's a final resolution.

**Comment by Mashiro2000** [comment]
This is my project code that solves the problem by ensuring the number of tasks never exceeds the number of subprocess handlers at any time. I use `asyncio.Semaphore` to limit the concurrent execution count.
```python
import asyncio
import inspect
import os
import threading
from asyncio import futures
from concurrent.futures import ProcessPoolExecutor
from typing import Optional, Dict, Callable

from loguru import logger


class ProcessPoolManager:
    _pools: Dict[str, ProcessPoolExecutor] = {}
    _tasks: Dict[str, Callable] = {}
    _lock = threading.RLock()
    _async_lock = asyncio.Lock()
    _semaphores: Dict[str, asyncio.Semaphore] = {}

    @classmethod
    def get_executor(
            cls,
            name: str,
            max_workers: Optional[int] = None,
            max_tasks_per_child: Optional[int]=None
    ) -> ProcessPoolExecutor:
        with cls._lock:
            if name not in cls._pools:
                logger.info(f"Creating executor pool: {name}")
                max_workers = max_workers or os.cpu_count() - 1
                cls._pools[name] = ProcessPoolExecutor(max_workers=max_workers, max_tasks_per_child=max_tasks_per_child)
                if max_tasks_per_child:
                    cls._semaphores[name] = asyncio.Semaphore(max_tasks_per_child)
                else:
                    cls._semaphores[name] = asyncio.Semaphore(100)
            return cls._pools[name]

    @classmethod
    async def async_run_executor(
            cls,
            name: str,
            func: Callable,
            reference_func: Callable | None = None,
            *args,
            **kwargs,
    ):
        loop = asyncio.get_event_loop()
        async with cls._async_lock:
            executor = cls.get_executor(name=name)
        func_kwargs = {
            key: value for key, value in kwargs.items()
            if key in inspect.signature(func).parameters
        }
        if reference_func:
            func_kwargs.update({
                key: value for key, value in kwargs.items()
                if key in inspect.signature(reference_func).parameters
            })
        if name in cls._semaphores:
            async with cls._semaphores[name]:
                future = executor.submit(func, *args, **func_kwargs)
                return await futures.wrap_future(future, loop=loop)
        else:
            future = executor.submit(func, *args, **func_kwargs)
            return await futures.wrap_future(future, loop=loop)
```

**Comment by karolinepauls** [comment]
Would it be possible to add a warning not to use `max_tasks_per_child` to the documentation until the fix is introduced and backported?

**Comment by john970** [comment]
Just wanted to put in another vote to update the docs to indicate this does not work, or better, just make it a noop and a warning. Could save a lot of frustration and time for folks using this...

**Comment by yaogang2060** [comment]
```python
from concurrent.futures import ProcessPoolExecutor
if __name__ == '__main__':
    with ProcessPoolExecutor(3, max_tasks_per_child=1) as exe:
        futs = [exe.submit(print, i) for i in range(10)]
```
set `max_tasks_per_child` to 1 works.

**Comment by Mayuresh16** [comment]
@gaogaotiantian @gpshead  Can we have a fix for this long-standing bug in the upcoming Python 3.15?

**Comment by jokester** [comment]
before this can be fixed i'm using https://loky.readthedocs.io/en/stable/ for a safer process worker pool

**Comment by csm10495** [comment]
Any chance this could be fixed soon? There are a couple PRs that intend to resolve it already.

I just bumped into seeing this issue when trying to use this feature for the first time. Now I'm a bit fearful to use it since I'm not sure if it'll hang.

**Comment by gpshead** [comment]
yep! I finally picked that PR back up. Updated, just waiting on CI to pass.
